### PR TITLE
Remove the usage of string refs

### LINKED ISSUE #2248: Cannot read property 'refs' of undefined
Happened randomly when running locally.

image

### LINKED ISSUE #2712: Remove usage of string refs
String refs in React are deprecated / buggy and shouldn't be used.

### PR DESCRIPTION
**Pull request by yuqijin**
### Reasons for making this change

Remove the usage of string refs

### Related issues

fixes #2712 
fixes #2248

### CODE REVIEW

File: frontend/src/components/worksheets/Worksheet/Worksheet.js [hunk 1]
```diff
@@ -673,22 +673,21 @@ class Worksheet extends React.Component {
...
                 let screenScrollTop = $(window).scrollTop();
                 let screenHeight = $(window).height();
-                let scrollIsAboveElement = elementOffsetTop + elementHeight - screenScrollTop >= 0;
+                // 170 is the sum of height of WorksheetHeader and NavBar.
+                // Since they block the user's view, we should take them into account when calculating whether the item is on the screen or not.
+                let scrollIsAboveElement =
+                    elementOffsetTop + elementHeight - screenScrollTop - 170 >= 0;
                 let elementIsVisibleOnScreen =
                     screenScrollTop + screenHeight - elementOffsetTop >= 0;
+                console.log(
```
Comment (epicfaace) [inline frontend/src/components/worksheets/Worksheet/Worksheet.js]: Can you remove this log statement?

File: frontend/src/components/worksheets/Worksheet/Worksheet.js [hunk 2]
```diff
@@ -673,22 +673,21 @@ class Worksheet extends React.Component {
                 let elementHeight = element.height();
                 let screenScrollTop = $(window).scrollTop();
                 let screenHeight = $(window).height();
-                let scrollIsAboveElement = elementOffsetTop + elementHeight - screenScrollTop >= 0;
+                // 170 is the sum of height of WorksheetHeader and NavBar.
+                // Since they block the user's view, we should take them into account when calculating whether the item is on the screen or not.
+                let scrollIsAboveElement =
+                    elementOffsetTop + elementHeight - screenScrollTop - 170 >= 0;
```
Comment (epicfaace) [inline frontend/src/components/worksheets/Worksheet/Worksheet.js]: Does this change have to do with this PR ("remove the usage of string refs")?

File: frontend/src/components/worksheets/Worksheet/Worksheet.js (hunk 1 above)
Comment (yuqijin) [reply frontend/src/components/worksheets/Worksheet/Worksheet.js]: Sure, I have deleted it.

File: frontend/src/components/worksheets/Worksheet/Worksheet.js (hunk 2 above)
Comment (yuqijin) [reply frontend/src/components/worksheets/Worksheet/Worksheet.js]: Not exactly. I just created a new PR #3023 to make it clearer!

File: frontend/src/components/worksheets/Worksheet/Worksheet.js [hunk 3]
```diff
@@ -680,15 +680,7 @@ class Worksheet extends React.Component {
             }
             shouldScroll = !isOnScreen(element);
         }
-        if (item && (!item.props || !item.props.item)) {
```
Comment (epicfaace) [inline frontend/src/components/worksheets/Worksheet/Worksheet.js:684]: Is this piece of code no longer needed?

Comment (yuqijin) [reply frontend/src/components/worksheets/Worksheet/Worksheet.js:684]: You can easily find there are no refs of children mode for worksheetItemList component, so the variable 'item' here is always undefined, which causes the line never be executed.

**Review by epicfaace — APPROVED** [review]

**Pull request merged**